src/uucore/src/lib/features/sum.rs

```diff
@@ -0,0 +1,486 @@
+            crc_table: Self::generate_crc_table(),
+        }
+    }
+
+    fn hash_update(&mut self, input: &[u8]) {
+        for &elt in input.iter() {
+            self.update(elt);
+        }
+        self.size += input.len();
+    }
+
+    fn hash_finalize(&mut self, out: &mut [u8]) {
+        let mut sz = self.size;
+        while sz != 0 {
+            self.update(sz as u8);
+            sz >>= 8;
+        }
+        self.state = !self.state;
+        out.copy_from_slice(&self.state.to_ne_bytes());
+    }
+
+    fn result_str(&mut self) -> String {
+        let mut _out: Vec<u8> = vec![0; 4];
+        self.hash_finalize(&mut _out);
+        format!("{}", self.state)
+    }
+
+    fn reset(&mut self) {
+        *self = Self::new();
+    }
+
+    fn output_bits(&self) -> usize {
+        256
+    }
+}
+
+// This can be replaced with usize::div_ceil once it is stabilized.
+// This implementation approach is optimized for when `b` is a constant,
+// particularly a power of two.
+pub fn div_ceil(a: usize, b: usize) -> usize {
+    (a + b - 1) / b
+}
+
+pub struct BSD {
+    state: u16,
+}
+impl Digest for BSD {
+    fn new() -> Self {
+        Self { state: 0 }
+    }
+
+    fn hash_update(&mut self, input: &[u8]) {
+        for &byte in input.iter() {
+            self.state = (self.state >> 1) + ((self.state & 1) << 15);
+            self.state = self.state.wrapping_add(u16::from(byte));
+        }
+    }
+
+    fn hash_finalize(&mut self, out: &mut [u8]) {
+        out.copy_from_slice(&self.state.to_ne_bytes());
+    }
+
+    fn result_str(&mut self) -> String {
+        let mut _out: Vec<u8> = vec![0; 2];
+        self.hash_finalize(&mut _out);
+        format!("{}", self.state)
+    }
+
+    fn reset(&mut self) {
+        *self = Self::new();
+    }
+
+    fn output_bits(&self) -> usize {
+        128
+    }
+}
+
+pub struct SYSV {
+    state: u32,
+}
+impl Digest for SYSV {
+    fn new() -> Self {
+        Self { state: 0 }
+    }
+
+    fn hash_update(&mut self, input: &[u8]) {
+        for &byte in input.iter() {
+            self.state = self.state.wrapping_add(u32::from(byte));
+        }
+    }
+
+    fn hash_finalize(&mut self, out: &mut [u8]) {
+        self.state = (self.state & 0xffff) + (self.state >> 16);
+        self.state = (self.state & 0xffff) + (self.state >> 16);
+        out.copy_from_slice(&(self.state as u16).to_ne_bytes());
+    }
+
+    fn result_str(&mut self) -> String {
+        let mut _out: Vec<u8> = vec![0; 2];
+        self.hash_finalize(&mut _out);
+        format!("{}", self.state)
+    }
+
+    fn reset(&mut self) {
+        *self = Self::new();
+    }
+
+    fn output_bits(&self) -> usize {
+        512
+    }
+}
+
+// Implements the Digest trait for sha2 / sha3 algorithms with fixed output
+macro_rules! impl_digest_common {
+    ($algo_type: ty, $ori_algo_type: ty, $size: expr) => {
+        impl Digest for $algo_type {
+            fn new() -> Self {
+                Self(<$ori_algo_type>::default())
```

Comment:
I think we can remove the `ori_algo_type` parameter if we use `Default::default()` here.

Reply:
done